flake8 errors on github actions

* glotaran version: master branch
* Python version: github actions flake8 checking

```
.\glotaran\analysis\matrix_calculation.py:50:74: E741 ambiguous variable name 'l'
.\glotaran\analysis\residual_calculation.py:84:5: E306 expected 1 blank line before a nested definition, found 0
.\glotaran\analysis\simulation.py:10:33: F821 undefined name 'glotaran'
.\glotaran\analysis\simulation.py:74:34: F541 f-string is missing placeholders
.\glotaran\builtin\models\doas\doas_matrix.py:18:42: E741 ambiguous variable name 'l'
.\glotaran\builtin\models\doas\doas_matrix.py:18:72: E741 ambiguous variable name 'l'
.\glotaran\builtin\models\kinetic_image\kinetic_image_matrix.py:128:1: E302 expected 2 blank lines, found 0
.\glotaran\builtin\models\kinetic_spectrum\kinetic_spectrum_model.py:21:54: F821 undefined name 'KineticModel'
.\glotaran\builtin\models\kinetic_spectrum\kinetic_spectrum_model.py:26:28: F821 undefined name 'KineticModel'
.\glotaran\builtin\models\kinetic_spectrum\kinetic_spectrum_model.py:36:40: F821 undefined name 'KineticModel'
.\glotaran\builtin\models\kinetic_spectrum\kinetic_spectrum_model.py:49:32: F821 undefined name 'KineticModel'
.\glotaran\builtin\models\kinetic_spectrum\spectral_constraints.py:82:28: F821 undefined name 'KineticModel'
.\glotaran\builtin\models\kinetic_spectrum\spectral_penalties.py:45:47: F821 undefined name 'KineticModel'
.\glotaran\builtin\models\kinetic_spectrum\spectral_penalties.py:50:28: F821 undefined name 'KineticModel'
.\glotaran\builtin\models\kinetic_spectrum\spectral_relations.py:35:28: F821 undefined name 'KineticModel'
.\glotaran\builtin\models\kinetic_spectrum\spectral_relations.py:57:28: F821 undefined name 'KineticModel'
.\glotaran\builtin\models\kinetic_spectrum\spectral_relations.py:75:28: F821 undefined name 'KineticModel'
.\glotaran\cli\commands\explore.py:6:1: E302 expected 2 blank lines, found 1
.\glotaran\cli\commands\optimize.py:92:32: F541 f-string is missing placeholders
.\glotaran\parameter\parameter_group.py:324:17: E741 ambiguous variable name 'l'
.\glotaran\parameter\parameter_group.py:358:13: E741 ambiguous variable name 'l'
```

--- a/glotaran/analysis/matrix_calculation.py
+++ b/glotaran/analysis/matrix_calculation.py
@@ -47,7 +47,7 @@ def calculate_index_independend_grouped_matrices(scheme, groups, parameter):
     for label, group in groups.items():
         if label not in matrices:
             constraint_labels_and_matrices[label] = \
-                _combine_matrices([constraint_labels_and_matrices[l] for l in group])
+                _combine_matrices([constraint_labels_and_matrices[label] for label in group])
 
     return clp_labels, matrices, constraint_labels_and_matrices
 

--- a/glotaran/analysis/residual_calculation.py
+++ b/glotaran/analysis/residual_calculation.py
@@ -79,7 +79,7 @@ def create_index_independend_grouped_residual(
         scheme, parameter, problem_bag, constraint_labels_and_matrices, residual_function):
 
     matrix_labels = problem_bag.pluck(1)\
-        .map(lambda group: "".join(problem.dataset for problem in group))\
+        .map(lambda group: "".join(problem.dataset for problem in group))
 
     def penalty_function(matrix_label, problem, labels_and_matrices):
 

--- a/glotaran/analysis/simulation.py
+++ b/glotaran/analysis/simulation.py
@@ -6,8 +6,10 @@ import xarray as xr
 
 from glotaran.parameter import ParameterGroup
 
+T_Model = typing.TypeVar('glotaran.model.Model')
 
-def simulate(model: typing.Type['glotaran.model.Model'],
+
+def simulate(model: T_Model,
              dataset: str,
              parameter: ParameterGroup,
              axes: typing.Dict[str, np.ndarray] = None,
@@ -71,7 +73,7 @@ def simulate(model: typing.Type['glotaran.model.Model'],
             if model.global_dimension not in clp.coords:
                 raise ValueError(f"Missing coordinate '{model.global_dimension}' in clp.")
             if 'clp_label' not in clp.coords:
-                raise ValueError(f"Missing coordinate 'clp_label' in clp.")
+                raise ValueError("Missing coordinate 'clp_label' in clp.")
         else:
             if 'clp_label' not in axes:
                 raise ValueError("Missing axis 'clp_label'")

--- a/glotaran/builtin/models/doas/doas_matrix.py
+++ b/glotaran/builtin/models/doas/doas_matrix.py
@@ -14,8 +14,10 @@ def calculate_doas_matrix(dataset_descriptor=None, axis=None, index=None, irf=No
     oscillations = _collect_oscillations(dataset_descriptor)
     matrix = np.zeros((axis.size, len(2 * oscillations)), dtype=np.float64)
     labels = [o.label for o in oscillations]
+    # TODO: improve code reasability further
     clp = [
-        lbl for o in zip([f'{l}_cos' for l in labels], [f'{l}_sin' for l in labels]) for lbl in o
+        lbl for o in zip([f'{l_cos}_cos' for l_cos in labels],
+                         [f'{l_sin}_sin' for l_sin in labels]) for lbl in o
     ]
 
     delta = np.abs(axis[1:] - axis[:-1])

--- a/glotaran/builtin/models/kinetic_image/kinetic_image_matrix.py
+++ b/glotaran/builtin/models/kinetic_image/kinetic_image_matrix.py
@@ -5,6 +5,8 @@ import numba as nb
 
 from .irf import IrfMultiGaussian
 
+sqrt2 = np.sqrt(2)
+
 
 def kinetic_image_matrix(
         dataset_descriptor=None, axis=None, index=None, irf=None):
@@ -124,7 +126,6 @@ def calculate_kinetic_matrix_no_irf(matrix, rates, times):
             matrix[n_t, n_r] += np.exp(r_n * t_n)
 
 
-sqrt2 = np.sqrt(2)
 @nb.jit(nopython=True, parallel=True)
 def calculate_kinetic_matrix_gaussian_irf(
         matrix, rates, times, center, width, scale, backsweep, backsweep_period):

--- a/glotaran/builtin/models/kinetic_image/kinetic_image_result.py
+++ b/glotaran/builtin/models/kinetic_image/kinetic_image_result.py
@@ -2,14 +2,16 @@ import typing
 import numpy as np
 import xarray as xr
 
-import glotaran
 from glotaran.parameter import ParameterGroup
 
 from .irf import IrfMultiGaussian
 
+T_KineticImageModel = typing.TypeVar(
+    'glotaran.builtin.models.kinetic_image.KineticImageModel')
+
 
 def finalize_kinetic_image_result(
-    model: 'glotaran.models.spectral_temporal.KineticModel',
+    model: T_KineticImageModel,
     global_indices: typing.List[typing.List[object]],
     reduced_clp_labels: typing.Union[typing.Dict[str, typing.List[str]], np.ndarray],
     reduced_clps: typing.Union[typing.Dict[str, np.ndarray], np.ndarray],

--- a/glotaran/builtin/models/kinetic_spectrum/kinetic_spectrum_model.py
+++ b/glotaran/builtin/models/kinetic_spectrum/kinetic_spectrum_model.py
@@ -17,13 +17,16 @@ from .spectral_penalties import EqualAreaPenalty, has_spectral_penalties, apply_
 from .spectral_relations import SpectralRelation, apply_spectral_relations
 from .spectral_shape import SpectralShape
 
+T_KineticSpectrumModel = typing.TypeVar(
+    'glotaran.builtin.models.kinetic_spectrum.KineticSpectrumModel')
 
-def has_kinetic_model_constraints(model: typing.Type['KineticModel']) -> bool:
+
+def has_kinetic_model_constraints(model: T_KineticSpectrumModel) -> bool:
     return len(model.spectral_relations) + len(model.spectral_constraints) != 0
 
 
 def apply_kinetic_model_constraints(
-        model: typing.Type['KineticModel'],
+        model: T_KineticSpectrumModel,
         parameter: ParameterGroup,
         clp_labels: typing.List[str],
         matrix: np.ndarray,
@@ -33,7 +36,7 @@ def apply_kinetic_model_constraints(
     return clp_labels, matrix
 
 
-def index_dependend(model: typing.Type['KineticModel']):
+def index_dependend(model: T_KineticSpectrumModel):
     if any([
         isinstance(irf, IrfSpectralMultiGaussian) and irf.dispersion_center is not None
         for irf in model.irf.values()
@@ -46,7 +49,7 @@ def index_dependend(model: typing.Type['KineticModel']):
     return False
 
 
-def grouped(model: typing.Type['KineticModel']):
+def grouped(model: T_KineticSpectrumModel):
     return len(model.dataset) != 1
 
 

--- a/glotaran/builtin/models/kinetic_spectrum/kinetic_spectrum_result.py
+++ b/glotaran/builtin/models/kinetic_spectrum/kinetic_spectrum_result.py
@@ -1,22 +1,24 @@
 import typing
+
 import numpy as np
 import xarray as xr
 
-import glotaran
-from glotaran.parameter import ParameterGroup
 from glotaran.builtin.models.kinetic_image.irf import IrfMultiGaussian
 from glotaran.builtin.models.kinetic_image.kinetic_image_result import (
-    retrieve_species_assocatiated_data,
-    retrieve_decay_assocatiated_data,
-    retrieve_irf,
-)
+    retrieve_decay_assocatiated_data, retrieve_irf,
+    retrieve_species_assocatiated_data)
+from glotaran.parameter import ParameterGroup
 
-from .spectral_irf import IrfSpectralMultiGaussian, IrfGaussianCoherentArtifact
 from .spectral_constraints import OnlyConstraint, ZeroConstraint
+from .spectral_irf import IrfGaussianCoherentArtifact, IrfSpectralMultiGaussian
+
+
+T_KineticSpectrumModel = typing.TypeVar(
+    'glotaran.builtin.models.kinetic_spectrum.KineticSpectrumModel')
 
 
 def finalize_kinetic_spectrum_result(
-    model: 'glotaran.models.spectral_temporal.KineticModel',
+    model: T_KineticSpectrumModel,
     global_indices: typing.List[typing.List[object]],
     reduced_clp_labels: typing.Union[typing.Dict[str, typing.List[str]], np.ndarray],
     reduced_clps: typing.Union[typing.Dict[str, np.ndarray], np.ndarray],

--- a/glotaran/builtin/models/kinetic_spectrum/spectral_constraints.py
+++ b/glotaran/builtin/models/kinetic_spectrum/spectral_constraints.py
@@ -1,9 +1,13 @@
 """This package contains compartment constraint items."""
 
 import typing
+
 import numpy as np
 
 from glotaran.model import model_attribute, model_attribute_typed
+
+T_KineticSpectrumModel = typing.TypeVar(
+    'glotaran.builtin.models.kinetic_spectrum.KineticSpectrumModel')
 
 
 @model_attribute(
@@ -79,7 +83,7 @@ class SpectralConstraint:
 
 
 def apply_spectral_constraints(
-        model: typing.Type['KineticModel'],
+        model: T_KineticSpectrumModel,
         clp_labels: typing.List[str],
         matrix: np.ndarray,
         index: float) -> typing.Tuple[typing.List[str], np.ndarray]:

--- a/glotaran/builtin/models/kinetic_spectrum/spectral_penalties.py
+++ b/glotaran/builtin/models/kinetic_spectrum/spectral_penalties.py
@@ -8,6 +8,9 @@ from glotaran.parameter import Parameter, ParameterGroup
 
 from .spectral_relations import retrieve_clps
 
+T_KineticSpectrumModel = typing.TypeVar(
+    'glotaran.builtin.models.kinetic_spectrum.KineticSpectrumModel')
+
 
 @model_attribute(properties={
     'compartment': str,
@@ -42,12 +45,12 @@ class EqualAreaPenalty:
         return any([applies(i) for i in self.interval])
 
 
-def has_spectral_penalties(model: typing.Type['KineticModel']) -> bool:
+def has_spectral_penalties(model: T_KineticSpectrumModel) -> bool:
     return len(model.equal_area_penalties) != 0
 
 
 def apply_spectral_penalties(
-        model: typing.Type['KineticModel'],
+        model: T_KineticSpectrumModel,
         parameter: ParameterGroup,
         clp_labels: typing.List[str],
         clps: np.ndarray,

--- a/glotaran/builtin/models/kinetic_spectrum/spectral_relations.py
+++ b/glotaran/builtin/models/kinetic_spectrum/spectral_relations.py
@@ -1,10 +1,14 @@
 """ Glotaran Spectral Relation """
 
 import typing
+
 import numpy as np
 
 from glotaran.model import model_attribute
 from glotaran.parameter import Parameter, ParameterGroup
+
+T_KineticSpectrumModel = typing.TypeVar(
+    'glotaran.builtin.models.kinetic_spectrum.KineticSpectrumModel')
 
 
 @model_attribute(
@@ -32,7 +36,7 @@ class SpectralRelation:
 
 
 def create_spectral_relation_matrix(
-        model: typing.Type['KineticModel'],
+        model: T_KineticSpectrumModel,
         parameter: ParameterGroup,
         clp_labels: typing.List[str],
         matrix: np.ndarray,
@@ -54,7 +58,7 @@ def create_spectral_relation_matrix(
 
 
 def apply_spectral_relations(
-        model: typing.Type['KineticModel'],
+        model: T_KineticSpectrumModel,
         parameter: ParameterGroup,
         clp_labels: typing.List[str],
         matrix: np.ndarray,
@@ -72,7 +76,7 @@ def apply_spectral_relations(
 
 
 def retrieve_clps(
-        model: typing.Type['KineticModel'],
+        model: T_KineticSpectrumModel,
         parameter: ParameterGroup,
         reduced_clp_labels: typing.List[str],
         reduced_clps: np.ndarray,

--- a/glotaran/cli/commands/explore.py
+++ b/glotaran/cli/commands/explore.py
@@ -3,6 +3,7 @@ from click import echo, pause, prompt
 
 from . import util
 
+
 @click.option('--name', '-n', default=None, type=str, show_default=True,
               help='Name of the datagroup to export.')
 @click.option('--select', '-s', default=None, type=(str, util.VALORRANGEORLIST), show_default=True,

--- a/glotaran/cli/commands/optimize.py
+++ b/glotaran/cli/commands/optimize.py
@@ -89,7 +89,7 @@ def optimize_cmd(dataformat: str, data: typing.List[str], out: str, nfev: int, n
                 click.echo(f"Saving directory is '{out}'")
                 if yes or click.confirm('Do you want to save the data?', default=True):
                     paths = result.save(out)
-                    click.echo(f"File saving successfull, the follwing files have been written:\n")
+                    click.echo("File saving successfull, the follwing files have been written:\n")
                     for p in paths:
                         click.echo(f"* {p}")
             except Exception as e:

--- a/glotaran/model/model.py
+++ b/glotaran/model/model.py
@@ -4,6 +4,7 @@
 import copy
 import inspect
 import typing
+
 import numpy as np
 import xarray as xr
 

--- a/glotaran/parameter/parameter_group.py
+++ b/glotaran/parameter/parameter_group.py
@@ -321,8 +321,8 @@ class ParameterGroup(dict):
     def groups(self) -> typing.Generator['ParameterGroup', None, None]:
         """Returns a generator over all groups and their subgroups."""
         for group in self:
-            for l in group.groups():
-                yield l
+            for sub_group in group.groups():
+                yield sub_group
 
     def has(self, label: str) -> bool:
         """Checks if a parameter with the given label is in the group or in a subgroup.
@@ -354,10 +354,11 @@ class ParameterGroup(dict):
         path = label.split(".")
         label = path.pop()
 
+        # TODO: audit this code
         group = self
-        for l in path:
+        for element in path:
             try:
-                group = group[l]
+                group = group[element]
             except KeyError:
                 raise ParameterNotFoundException(path, label)
         try:

